### DISCUSSION TITLE
How to customize the sticky note when you use `add_text_annot()`?

### DISCUSSION THREAD

**Discussion by danejo3** [pymupdf/PyMuPDF#1189] [Q&A]
Hi everyone!

Is there a way to edit/update a sticky note's fill and size and the inside text's font and size?

Updating the `annot` object created by `add_text_annot()` will only update the icon, not the popup.

In the documention, I see that we have `add_freetext_annot()` to do all the following; however, I would like to somehow replace or update the sticky note with this new `annot` object to preserve the "popup/hover" feature.

Any suggestions would be greatly appreciated! Thanks!

Also, is there a way to insert a string when you call `annot.set_popup()`? The function only takes in a `rect` object and with it, it creates an empty popup. I do not understand the purpose of this function if we cannot put anything inside the popup; unless, this function was intended to create a popup for the users to fill in manually.

**Comment by JorjMcKie** [comment]
> Updating the annot object created by add_text_annot() will only update the icon, not the popup.

This annot type has no fill color. The color you see in the icon is the stroke color - which can be changed of course.

> this function was intended to create a popup for the users to fill in manually.

This indeed is the intention. This, and / or set a not popup rect position. A user updating the text in the popup will thereby change the `"content"` of the **_original_** annot.
You can also set and inquire the popup's "open" state.

**Reply by danejo3** [reply]
Thanks @JorjMcKie for the quick response! Just to clarify, when you create a sticky note with `add_text_annot()`, there is no way to adjust the size and color of the sticky note and the text's font and size inside it, correct?

**Reply by JorjMcKie** [reply]
> there is no way to adjust the size and color of the sticky note and the text's font and size inside it, correct?

**_Not_** correct: you change that annot completely: color, location and size of the icon, the icon itself and the text hiding behing it. The font cannot be changed, but is irrelevant anyway, because the text is not displayed and only shows when hovering the mouse over the icon.

**Reply by JorjMcKie** [reply]
The following 3 text annots are e.g. possible:
![grafik](https://user-images.githubusercontent.com/8290722/128357531-d27fa270-2f82-4275-a62c-244604d9e30e.png)

**Reply by JorjMcKie** [reply]
The font itself (as shown in any popup) can indeed not be influenced.

**Reply by JorjMcKie** [reply]
Of course you can update the annot text font using some other tool. But as soon as you update the annot again with PyMuPDF, it will again be reset.
The same or similar will BTW probably also happen, if you change with a different tool - they all have their own capabilities / features, and there is no general prescription / spec for how to behave in that respect.

**Reply by tyang209** [reply]
@JorjMcKie First off I must say this is an incredible PDF library! Incredible documentation and amazing support from you.

I saw this comment and am trying to emulate it. I'm struggling to scale a text annot properly after creating it.



``` python
pos = (10, 10)
annot = page.add_text_annot(positions[x], "Test Definition", icon="Help")
r = annot.rect
r.x1 = r.x0 + 10 # shrink it by half since default is 20x20
r.y1 = r.y0 + 10
annot.set_rect(r)
```

This doesn't seem to affect the size in Preview. Is it possible that it's a Program specific issue? Thank you so much for looking at this if you do at all! Really appreciate the package

**Reply by JorjMcKie** [reply]
First off: thank you very much for your appreciation!

> This doesn't seem to affect the size in Preview.

Depends a little what "Preview" really is. This at least works no problem:
```python
doc = fitz.open()
page = doc.new_page()
a1 = page.add_text_annot((100, 100), "text 1")
a2 = page.add_text_annot(a1.rect.br, "text 2")
clip = a1.rect | a2.rect
r = a2.rect
r.x1 -= 10
r.y1 -= 10
a2.set_rect(r)
pix = page.get_pixmap(clip=clip)
pix.save("pic.png")
```
![pic](https://user-images.githubusercontent.com/8290722/156725753-35c49fe4-5305-4872-9d28-803d3840f33e.png)

But maybe you need to "reload" the page before you take pictures of your recent changes. This is advisable anyway after annot updates: because of my additions to the base library MuPDF, there sometimes is a "delay" before all the effects are applied: `page = doc.reload(page)`.

**Comment by RyanBerger98** [comment]
Hello, just to add some further clarification. I attached a screenshot of what exactly we are trying to change. 
<img width="582" alt="Screen Shot 2021-08-05 at 12 19 41 PM" src="https://user-images.githubusercontent.com/36561338/128387156-bafc7f04-b6f0-4fc2-ae5c-53dd99392c9c.png">

Our main question is if we are able to change the font and size of the text contained within the popup and if we are able to change the size of the popup? Not the icon responsible for the popup but the popup itself (outlined in red). We are also currently displaying the pdf in a browser (i.e Chrome), if that affects any capabilities.

**Reply by JorjMcKie** [reply]
* You **_cannot_** change the font and fontsize of the popup.
* You **_can_** set the popup rectangle (size & position) - see docu. If not existing, a popup annot will be created by this.
* You **_can_** set the initial open/close state of an annotation's popup.

In general, the annot popup display in any viewer is highly dependent on that viewer. I see no way to enforce a uniform behaviour across the myriad of viewers WRT this.
I  - decidedly - am no Chrome user, so won't be able to say what behaviour can be observed there. However, seems to be similar to MS Edge, which I have installed ... involuntarily. Looks like your picture - obviously b/o Edge is a Chrome clone.
Same thing looks completely different under Mozilla Firefox (my favorite browser).

As I previously wrote:
Except for the mentioned, the annot properties can be modified in their entirety.

**Reply by danejo3** [reply]
Perfect! Thank you @JorjMcKie! This answers my question. Thanks for creating this awesome package and maintaining such great documentation! 👏 👏 👏

**Reply by mst803** [reply]
Can you please clarify how to set the size of the popup rectangle with python code, since the set_popup method is not working....
i can share my code here

rect =fitz.Rect(20,20,40,40)
annoter = page.add_text_annot([rect.x1 -20, rect.y1 - 20],"Text Here")
annoter.set_popup(fitz.Rect(10.0, 10.0, 20.0, 20.0))
annoter.set_opacity(0)
annoter.update()
print(annoter.popup_rect)

**Comment by JorjMcKie** [comment]
... cont:
The first popup reflects the original annot's `"content"`. If there is a longer chain of popups for the same original, I don't know for sure ... yet, where their text would be stored.

Seems this is the `/RC` key of the resp. popup. There is no high level PyMuDF support to access this.
But you can use low-level code arbitrarily and parse your way through the jungle ... 😎
